# 1) Keyboard Shortcut List (Mac)

## Command Mode
- H show all shortcuts
- Ctrl + S save and checkpoint
- Enter take you into edit mode
- 1: change cell to heading 1
- 2: change cell to heading 2
- 3: change cell to heading 3
- 4: change cell to heading 4
- 5: change cell to heading 5
- 6: change cell to heading 6
- shift + enter run cell, select below
- ctrl + enter run cell
- option + enter run cell, insert below
- A insert cell above
- B insert cell below
- C copy cell
- V paste cell
- D , D delete selected cell
- shift + M merge selected cells, or current cell with cell below if only one cell selected
- I , I interrupt kernel
- 0 , 0 restart kernel (with dialog)
- Y change cell to code mode
- M change cell to markdown mode (good for documentation)


## Edit Mode
- cmd + click for multi-cursor editing
- option + scrolling click for column editing
- cmd + / toggle comment lines
- tab code completion or indent
- shift + tab tooltip
- ctrl + shift + - split cell

## Command Palette
- cmd + shift + p

## Sources: 
http://maxmelnick.com/2016/04/19/python-beginner-tips-and-tricks.html
https://towardsdatascience.com/jypyter-notebook-shortcuts-bf0101a98330

# 2) Text Formatting

# Heading 1 using "#"

## Heading 2 using "##"

### Heading 3 using "###"

`Adding a light background using` using ` 

- / * for bullet points

**Bold the words** using **

>Making Quotes using

| Make | Table | Using | Bamboo Stick |

In [8]:
# Enable Codes to run line by line
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# 3) Python Variables

In [36]:
#Creating Variables
x = 4 # x is of type int
y = "Sally" # x is now of type str
print('x: '+ str(x))
print('y: '+ y)


#Legal variable names:
myvar = "John"
my_var = "John"
_my_var = "John"
myVar = "John"
MYVAR = "John"
myvar2 = "John"


# Illegal variable names:
# 2myvar = "John"
# my-var = "John"
# my var = "John"


#Assign Value to Multiple Variables
x1, y1, z1 = "Orange", "Banana", "Cherry"
print('x1: '+ x1)
print('y1: '+ y1)
print('z1: '+ z1)


#Assign the same value to multiple variables in one line:
x2 = y2 = z2 = "Orange"
print('x2: '+ x2)
print('y2: '+ y2)
print('z2: '+ z2)


#To combine both text and a variable, Python uses the + character:
x3 = "awesome"
print("Python is " + x3)


#use the + character to add a variable to another variable:
x4 = "Python is "
y4 = "awesome2"
z4 =  x4 + y4
print(z4)


#For numbers, the + character works as a mathematical operator:
x5 = 5
y5 = 10
print('x5 + y5 = ' + str(x5 + y5))


# Global Variables
# Variables that are created outside of a function (as in all of the examples above) are known as global variables.
# Global variables can be used by everyone, both inside of functions and outside.
x6 = "awesome3"

def myfunc():
  print("Python x6 is " + x6)

myfunc()


# If you create a variable with the same name inside a function, this variable will be local, and can only be used inside the function. The global variable with the same name will remain as it was, global and with the original value.
x7 = "awesome"

def myfunc():
  x7 = "fantastic"
  print("Inner Python is " + x7)

myfunc()

print("Outer Python is " + x7)


#To create a global variable inside a function, you can use the global keyword.
def myfunc():
  global x8
  x8 = "fantasticcc"

myfunc()

print("Global Python is " + x8)


#To change the value of a global variable inside a function, refer to the variable by using the global keyword:
x9 = "awesome"

def myfunc():
  global x9
  x9 = "123 fantastic"

myfunc()

print("Changed Global Python is " + x9)

x: 4
y: Sally
x1: Orange
y1: Banana
z1: Cherry
x2: Orange
y2: Orange
z2: Orange
Python is awesome
Python is awesome2
x5 + y5 = 15
Python x6 is awesome3
Inner Python is fantastic
Outer Python is awesome
Global Python is fantasticcc
Changed Global Python is 123 fantastic


# 4) Date Type & Conversion

In [ ]:
#string
x = "Hello World"
x = str("Hello World")
#type(x) to check type, example:
print(type(x))

#int
x = 20
x = int(20)

#float
x = 20.5
x = float(20.5)

#complex, Complex numbers are specified as <real part>+ <imaginary part>j
x = 1 + 2j
x = complex(1j)

#list
# A list is a mutable, ordered sequence of items. As such, it can be indexed, sliced, 
# and changed. Each element can be accessed using its position in the list.
xlist = ["apple", "banana", "cherry", "cherry"]
x = list(("apple", "banana", "cherry"))

#tuple
# Think of them as similar to lists, but without the extensive functionality that the list class gives you. 
# One major feature of tuples is that they are immutable like strings i.e. you cannot modify tuples. 
xturple = ("apple", "banana", "cherry", "cherry")
x = tuple(("apple", "banana", "cherry"))

#set
# Sets cannot have multiple occurrences of the same element, 
# it makes sets highly useful to efficiently remove duplicate values from a list or tuple 
# and to perform common math operations like unions and intersections.
xset = {"apple", "banana", "cherry", "cherry"}
x = set(("apple", "banana", "cherry"))
print(xset) #only display unit values



 Lists and tuples are standard Python data types that store values in a sequence. <br>
 Sets are another standard Python data type that also store values. <br>
 The major difference is that sets, unlike lists or tuples, <br>
 cannot have multiple occurrences of the same element and store unordered values. <br>
 The main difference between lists and tuples is the fact that <br>
 Lists are mutable whereas tuples are immutable. <br>
 A mutable data type means that a python object of this type can be modified. <br>
 An immutable object can’t.

In [46]:
#dict, key value pairs
#A dictionary is like an address-book where you can find the address or contact details of a person by knowing only his/her name i.e. 
#We associate keys (name) with values (details). 
#Note that the key must be unique just like you cannot find out the correct information if you have two persons with the exact same name.
x = {"name" : "John", "age" : 36}
x = dict(name="John", age=36)

#frozenset
#Frozen set is just an immutable version of a Python set object. While elements of a set can be modified at any time, elements of the frozen set remain the same after creation.
x = frozenset({"apple", "banana", "cherry"})

#range, The range() function is used to generate a sequence of numbers over time
x = range(6)

#bool
x = True
x = bool(5)

#bytes
xbytes = b"Hello"
x = bytes(5)

#bytearray
x = bytearray(5)

#memoryview
x = memoryview(bytes(5))

<class 'str'>
{'cherry', 'apple', 'banana'}


# 5) Python Numbers

There are three numeric types in Python: 
<br> int
<br> float
<br> complex

In [47]:
x = 1    # int, is a whole number, positive or negative, without decimals, of unlimited length.
y = 2.8  # float, is a number, positive or negative, containing one or more decimals.
z = 1j   # complex, are written with a "j" as the imaginary part:

#random number
import random
print(random.randrange(1, 10))

7


# 6) Python Casting
Specify a Variable Type
There may be times when you want to specify a type on to a variable. 
This can be done with casting. 
Python is an object-orientated language, 
and as such it uses classes to define data types, 
including its primitive types.

Casting in python is therefore done using constructor functions:

int() - constructs an integer number from an integer literal, a float literal (by rounding down to the previous whole number), or a string literal (providing the string represents a whole number)

float() - constructs a float number from an integer literal, a float literal or a string literal (providing the string represents a float or an integer)

str() - constructs a string from a wide variety of data types, including strings, integer literals and float literals

In [48]:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3

x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2

x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

# 7) Python Strings
String literals in python are surrounded by either single quotation marks, or double quotation marks.

'hello' is the same as "hello".

You can display a string literal with the print() function:

In [79]:
print("Hello")
print('Hello')

#Assign String to a Variable
a = "Hello"
print(a)

#Multiline Strings, You can assign a multiline string to a variable by using three quotes:
a1 = """Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua."""
print('a1 ' + a1)

#Or three single quotes:
a2 = '''Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.'''
print('a2 ' + a2)

#Strings are Arrays, Square brackets can be used to access elements of the string.
a3 = "Hello, World!"
print('a3 ' + a3[1])

#Slicing, Specify the start index and the end index, separated by a colon, to return a part of the string.
b1 = "Hello, World!"
print('b1 ' + b1[2:5])

#Negative Indexing, Use negative indexes to start the slice from the end of the string:
b2 = "Hello, World!"
print('b2 ' + b2[-5:-2])

#String Length, To get the length of a string, use the len() function.
a4 = "Hello, World!"
print('a4 ' + str(len(a4)))

#String Methods, 
#The strip() method removes any whitespace from the beginning or the end:
a5 = " Hello, World!     "
print('a5 ' + a5.strip()) # returns "Hello, World!"

#The lower() method returns the string in lower case:
a6 = "Hello, World!"
print('a6 ' + a6.lower())

#The upper() method returns the string in upper case:
a7 = "Hello, World!"
print('a7 ' + a7.upper())

#The replace() method replaces a string with another string: (First old, second New)
a8 = "Hello, World!"
print('a8 ' + a8.replace("H", "J"))

#The split() method splits the string into substrings if it finds instances of the separator:
a9 = "Hello, World!"
print(a9.split(",")) # returns ['Hello', ' World!']

# Check if the phrase "ain" is present in the following text:
txt = "The rain in Spain stays mainly in the plain"
a10 = "ain" in txt
print(a10)

# Check if the phrase "ain" is NOT present in the following text:
txt2 = "The rassin in Spain stays mainly in the plain"
a11 = "ain" not in txt2
print(a11) 

# String Concatenation, Merge variable a with variable b into variable c:
a = "Hello"
b = "World"
c = a + b
print(c)

#To add a space between them, add a " ":
a = "Hello"
b = "World"
c = a + " " + b
print(c)

#Use the format() method to insert numbers into strings
age = 36
number = 77
txt = "My name is John, and I am {}, and my favorite no is {}"
print(txt.format(age, number))

#You can use index numbers {0} to be sure the arguments are placed in the correct placeholders:
quantity = 3
itemno = 567
price = 49.95
myorder = "I want to pay {2} dollars for {0} pieces of item {1}."
print(myorder.format(quantity, itemno, price))

#The escape character allows you to use double quotes when you normally would not be allowed:
txt = "We are the so-called \"Vikings\" from the north."

Hello
Hello
Hello
a1 Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.
a2 Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.
a3 e
b1 llo
b2 orl
a4 13
a5 Hello, World!
a6 hello, world!
a7 HELLO, WORLD!
a8 Jello, World!
['Hello', ' World!']
True
False
HelloWorld
Hello World
My name is John, and I am 36, and my favorite no is 77
I want to pay 49.95 dollars for 3 pieces of item 567.


# 8) Python Booleans

In [ ]:
#Print a message based on whether the condition is True or False:
a = 200
b = 33

if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

#Print a message based on whether the condition is True or False:
print(bool("Hello"))
print(bool(15))

#Evaluate two variables:
x = "Hello"
y = 15

print(bool(x))
print(bool(y))

Most Values are True
Almost any value is evaluated to True if it has some sort of content.

Any string is True, except empty strings.

Any number is True, except 0.

Any list, tuple, set, and dictionary are True, except empty ones.

Some Values are False 

In fact, there are not many values that evaluates to False, except empty values, such as (), [], {}, "", the number 0, and the value None. 

And of course the value False evaluates to False.

In [80]:
#The following will return False:
bool(False)
bool(None)
bool(0)
bool("")
bool(())
bool([])
bool({})

False

False

False

False

False

False

False

One more value, or object in this case, evaluates to False, and that is if you have an object that is made from a class with a __len__ function that returns 0 or False:

In [81]:
class myclass():
  def __len__(self):
    return 0

myobj = myclass()
print(bool(myobj))

False


In [83]:
#Functions can Return a Boolean
def myFunction() :
  return True

print(myFunction())

def myFunction() :
  return True

if myFunction():
  print("YES!")
else:
  print("NO!")

True
YES!


Python also has many built-in functions that returns a boolean value, like the isinstance() function, which can be used to determine if an object is of a certain data type:

In [84]:
x = 200
print(isinstance(x, int))

True


# 9) Python Operators

Python divides the operators in the following groups:

Arithmetic operators <br>
Assignment operators <br>
Comparison operators <br>
Logical operators <br>
Identity operators <br>
Membership operators <br>
Bitwise operators

| Operator | Name |	Example	| 
| --- | --- | --- |
| +	| Addition | 	x + y	| 
| -	| Subtraction | x - y	| 
| *	| Multiplication	| x * y	| 
| /	| Division | x / y | 
| %	| Modulus | x % y	| 
| ** | Exponentiation |	x ** y	| 
| // | Floor division | x // y  | 

| Operator| 	Example| Same As |
| --- | --- | --- |
| = | x = 5	| x = 5	|
| +=	|x += 3	| x = x + 3 |
| -=	|x -= 3	| x = x - 3 |	
| *=	|x *= 3	| x = x * 3|	
| /=	|x /= 3	| x = x / 3|	
| %=	|x %= 3	| x = x % 3|	
| //=	|x //= 3	|x = x // 3|	
| **=	|x **= 3	|x = x ** 3|	
| &=	|x &= 3	| x = x & 3|	
|=	|x |= 3	| x = x | 3	|
|^=	| x ^= 3 |	x = x ^ 3	|
| >>= |	x >>= 3 | x = x >> 3	|
| <<= |	x <<= 3	| x = x << 3 |

|Operator	| Name	| Example	|
| --- | --- | --- |
|==	|Equal	|x == y	|
|!=	|Not equal	|x != y	|
|>	|Greater than|	x > y	|
|<	|Less than|x < y	|
|>=	|Greater than or equal to	|x >= y	|
|<=	|Less than or equal to	|x <= y|

|Operator	|Description|	Example	|
| --- | --- | --- |
|and 	|Returns True if both statements are true|	x < 5 and  x < 10	|
|or	|Returns True if one of the statements is true|	x < 5 or x < 4	|
|not|	Reverse the result, returns False if the result is true	| not(x < 5 and x < 10)|

|Operator|	Description|Example	|
| --- | --- | --- |
|is 	|Returns True if both variables are the same object|	x is y	|
|is not|	Returns True if both variables are not the same object	|x is not y|


|Operator	|Description|	Example|
| --- | --- | --- |
|in 	|Returns True if a sequence with the specified value is present in the object|	x in y|
|not in	|Returns True if a sequence with the specified value is not present in the object |x not in y|

|Operator	|Name	|Description|
| --- | --- | --- |
|& 	|AND	|Sets each bit to 1 if both bits are 1|
||	|OR	|Sets each bit to 1 if one of two bits is 1|
| ^	|XOR|	Sets each bit to 1 if only one of two bits is 1|
|~ 	|NOT|Inverts all the bits|
|<<	|Zero fill left shift|	Shift left by pushing zeros in from the right and let the leftmost bits fall off|
|>>	|Signed right shift|	Shift right by pushing copies of the leftmost bit in from the left, and let the rightmost bits fall off|

# 10) Python Collections (Arrays)

There are four collection data types in the Python programming language:

List is a collection which is ordered and changeable. Allows duplicate members. <br>
Tuple is a collection which is ordered and unchangeable. Allows duplicate members. <br>
Set is a collection which is unordered and unindexed. No duplicate members. <br>
Dictionary is a collection which is unordered, changeable and indexed. No duplicate members. <br>

When choosing a collection type, it is useful to understand the properties of that type. Choosing the right type for a particular data set could mean retention of meaning, and, it could mean an increase in efficiency or security.

## 10.1) List
A list is a collection which is ordered and changeable. In Python lists are written with square brackets.

In [92]:
# Create a List:
thislist = ["apple", "banana", "cherry"]
print(thislist)

# Print the second item of the list:
thislist = ["apple", "banana", "cherry"]
print(thislist[1])

#Negative indexing means beginning from the end, -1 refers to the last item, -2 refers to the second last item etc.
thislist = ["apple", "banana", "cherry"]
print(thislist[-1])

# Range of Indexes, Return the third, fourth, and fifth item:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[2:5])

#By leaving out the start value, the range will start at the first item:
#This example returns the items from the beginning to "orange":
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[:4])

# By leaving out the end value, the range will go on to the end of the list:
# This example returns the items from "cherry" and to the end:
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[2:])

# Specify negative indexes if you want to start the search from the end of the list:
# This example returns the items from index -4 (included) to index -1 (excluded)
thislist = ["apple", "banana", "cherry", "orange", "kiwi", "melon", "mango"]
print(thislist[-4:-1])

# Change Item Value
# To change the value of a specific item, refer to the index number:
thislist = ["apple", "banana", "cherry"]
thislist[1] = "blackcurrant"
print(thislist)

# Loop Through a List
thislist = ["apple", "banana", "cherry"]
for x in thislist:
  print(x)

# Check if Item Exists
# To determine if a specified item is present in a list use the in keyword:
thislist = ["apple", "banana", "cherry"]
if "apple" in thislist:
  print("Yes, 'apple' is in the fruits list")

# List Length
# To determine how many items a list has, use the len() function:
thislist = ["apple", "banana", "cherry"]
print(len(thislist))

# Add Items
# To add an item to the end of the list, use the append() method:
thislist = ["apple", "banana", "cherry"]
thislist.append("orange")
print(thislist)

# Insert an item as the second position:
thislist = ["apple", "banana", "cherry"]
thislist.insert(2, "orange")
print(thislist)

# The remove() method removes the specified item:
thislist = ["apple", "banana", "cherry"]
thislist.remove("banana")
print(thislist)

# The pop() method removes the specified index, (or the last item if index is not specified):
thislist = ["apple", "banana", "cherry"]
thislist.pop()
print(thislist)

# The del keyword removes the specified index:
thislist = ["apple", "banana", "cherry"]
del thislist[0]
print(thislist)

# The del keyword can also delete the list completely:
thislist = ["apple", "banana", "cherry"]
del thislist

# The clear() method empties the list:
thislist = ["apple", "banana", "cherry"]
thislist.clear()
print(thislist)

# Make a copy of a list with the copy() method:
thislist = ["apple", "banana", "cherry"]
mylist = thislist.copy()
print(mylist)

# Make a copy of a list with the list() method:
thislist = ["apple", "banana", "cherry"]
mylist = list(thislist)
print(mylist)

# Join Two Lists
list1 = ["a", "b" , "c"]
list2 = [1, 2, 3]

list3 = list1 + list2
print(list3)

# Append list2 into list1:
list1 = ["a", "b" , "c"]
list2 = [1, 2, 3]

for x in list2:
  list1.append(x)

print(list1)

# Use the extend() method to add list2 at the end of list1:
list1 = ["a", "b" , "c"]
list2 = [1, 2, 3]

list1.extend(list2)
print(list1)

# Using the list() constructor to make a List:
thislist = list(("apple", "banana", "cherry")) # note the double round-brackets
print(thislist)

['apple', 'banana', 'cherry']
banana
cherry
['cherry', 'orange', 'kiwi']
['apple', 'banana', 'cherry', 'orange']
['cherry', 'orange', 'kiwi', 'melon', 'mango']
['orange', 'kiwi', 'melon']
['apple', 'blackcurrant', 'cherry']
apple
banana
cherry
Yes, 'apple' is in the fruits list
3
['apple', 'banana', 'cherry', 'orange']
['apple', 'banana', 'orange', 'cherry']
['apple', 'cherry']


'cherry'

['apple', 'banana']
['banana', 'cherry']
[]
['apple', 'banana', 'cherry']
['apple', 'banana', 'cherry']
['a', 'b', 'c', 1, 2, 3]


| Method	| Description|
| ----------| -----------| 
| append()	| Adds an element at the end of the list| 
| clear()	| Removes all the elements from the list| 
| copy()	| Returns a copy of the list| 
| count()	| Returns the number of elements with the specified value| 
| extend()	| Add the elements of a list (or any iterable), to the end of the current list| 
| index()	| Returns the index of the first element with the specified value| 
| insert()	| Adds an element at the specified position| 
| pop()	| Removes the element at the specified position| 
| remove()	| Removes the item with the specified value| 
| reverse()	| Reverses the order of the list| 
| sort()	| Sorts the list| 


## 10.2) Python Tuples
A tuple is a collection which is ordered and unchangeable. In Python tuples are written with round brackets.

In [93]:
# Create a Tuple:
thistuple = ("apple", "banana", "cherry")
print(thistuple)

# Print the second item in the tuple:
thistuple = ("apple", "banana", "cherry")
print(thistuple[1])

# Print the last item of the tuple:
thistuple = ("apple", "banana", "cherry")
print(thistuple[-1])

# Return the third, fourth, and fifth item:
thistuple = ("apple", "banana", "cherry", "orange", "kiwi", "melon", "mango")
print(thistuple[2:5])

# This example returns the items from index -4 (included) to index -1 (excluded)
thistuple = ("apple", "banana", "cherry", "orange", "kiwi", "melon", "mango")
print(thistuple[-4:-1])

('apple', 'banana', 'cherry')
banana
cherry
('cherry', 'orange', 'kiwi')
('orange', 'kiwi', 'melon')


Change Tuple Values <br>
Once a tuple is created, you cannot change its values. 

Tuples are unchangeable, or immutable as it also is called.

But there is a workaround. You can convert the tuple into a list, change the list, and convert the list back into a tuple.

In [94]:
# Convert the tuple into a list to be able to change it:
x = ("apple", "banana", "cherry")
y = list(x)
y[1] = "kiwi"
x = tuple(y)

print(x)

('apple', 'kiwi', 'cherry')


In [95]:
# Iterate through the items and print the values:
thistuple = ("apple", "banana", "cherry")
for x in thistuple:
  print(x)

# Check if "apple" is present in the tuple:
thistuple = ("apple", "banana", "cherry")
if "apple" in thistuple:
  print("Yes, 'apple' is in the fruits tuple")

# Print the number of items in the tuple:
thistuple = ("apple", "banana", "cherry")
print(len(thistuple))


# Add Items
# Once a tuple is created, you cannot add items to it. Tuples are unchangeable.
# thistuple = ("apple", "banana", "cherry")
# thistuple[3] = "orange" # This will raise an error
# print(thistuple)

apple
banana
cherry
Yes, 'apple' is in the fruits tuple
3


In [96]:
# Create Tuple With One Item
# To create a tuple with only one item, you have to add a comma after the item, 
# otherwise Python will not recognize it as a tuple.
# One item tuple, remember the commma:

thistuple = ("apple",)
print(type(thistuple))

#NOT a tuple
thistuple = ("apple")
print(type(thistuple))

# The del keyword can delete the tuple completely:
thistuple = ("apple", "banana", "cherry")
del thistuple
# print(thistuple) #this will raise an error because the tuple no longer exists

# Join two tuples:
tuple1 = ("a", "b" , "c")
tuple2 = (1, 2, 3)

tuple3 = tuple1 + tuple2
print(tuple3)

# Using the tuple() method to make a tuple:
thistuple = tuple(("apple", "banana", "cherry")) # note the double round-brackets
print(thistuple)      

<class 'tuple'>
<class 'str'>
('a', 'b', 'c', 1, 2, 3)
('apple', 'banana', 'cherry')


|Method	| Description |
|-------|--------------|
|count()|Returns the number of times a specified value occurs in a tuple |
|index()|Searches the tuple for a specified value and returns the position of where it was found|

## 10.3) Python Sets
A set is a collection which is unordered and unindexed. In Python sets are written with curly brackets.

In [103]:
# Create a Set:
thisset = {"apple", "banana", "cherry"}
print(thisset)

# Loop through the set, and print the values:
thisset = {"apple", "banana", "cherry"}
for x in thisset:
  print(x)

# Check if "banana" is present in the set:
thisset = {"apple", "banana", "cherry"}
print("banana" in thisset)

# Add an item to a set, using the add() method:
thisset = {"apple", "banana", "cherry"}
thisset.add("orange")
print(thisset)

# Add multiple items to a set, using the update() method:
thisset = {"apple", "banana", "cherry"}
thisset.update(["orange", "mango", "grapes"])
print(thisset)

# Get the number of items in a set:
thisset = {"apple", "banana", "cherry"}
print(len(thisset))

# Remove "banana" by using the remove() method:
# Note: If the item to remove does not exist, remove() will raise an error.
thisset = {"apple", "banana", "cherry"}
thisset.remove("banana")
print(thisset)

# Remove "banana" by using the discard() method:
# Note: If the item to remove does not exist, discard() will NOT raise an error.
thisset = {"apple", "banana", "cherry"}
thisset.discard("banana")
print(thisset)

# Remove the last item by using the pop() method:
thisset = {"apple", "banana", "cherry"}
x = thisset.pop()
print(x)
print(thisset)

# The clear() method empties the set:
thisset = {"apple", "banana", "cherry"}
thisset.clear()
print(thisset)

# The del keyword will delete the set completely:
thisset = {"apple", "banana", "cherry"}
del thisset

# Both union() and update() will exclude any duplicate items.
# The union() method returns a new set with all items from both sets:
set1 = {"a", "b" , "c"}
set2 = {1, 2, 3}

set3 = set1.union(set2)
print(set3)

# The update() method inserts the items in set2 into set1:
set1 = {"a", "b" , "c"}
set2 = {1, 2, 3}

set1.update(set2)
print(set1)

# Using the set() constructor to make a set:
thisset = set(("apple", "banana", "cherry")) # note the double round-brackets
print(thisset)

{'cherry', 'apple', 'banana'}
cherry
apple
banana
True
{'cherry', 'orange', 'apple', 'banana'}
{'orange', 'apple', 'cherry', 'grapes', 'mango', 'banana'}
3
{'cherry', 'apple'}
{'cherry', 'apple'}
cherry
{'apple', 'banana'}
set()
{'c', 1, 2, 3, 'a', 'b'}
{'c', 1, 2, 3, 'a', 'b'}
{'cherry', 'apple', 'banana'}


|Method|	Description |
|-------|-----------|
|add()	|Adds an element to the set|
|clear()|	Removes all the elements from the set|
|copy()|	Returns a copy of the set|
|difference()|	Returns a set containing the difference between two or more sets|
|difference_update()	|Removes the items in this set that are also included in another, |specified set|
|discard()	|Remove the specified item|
|intersection()|	Returns a set, that is the intersection of two other sets|
|intersection_update()|	Removes the items in this set that are not present in other, |specified set(s)|
|isdisjoint()|	Returns whether two sets have a intersection or not|
|issubset()|	Returns whether another set contains this set or not|
|issuperset()|	Returns whether this set contains another set or not|
|pop()	|Removes an element from the set|
|remove()	|Removes the specified element|
|symmetric_difference()|	Returns a set with the symmetric differences of two sets|
|symmetric_difference_update()	|inserts the symmetric differences from this set and another|
|union()	|Return a set containing the union of sets|
|update()	|Update the set with the union of this set and others|

## 10.4) Python Dictionaries
A dictionary is a collection which is unordered, changeable and indexed. In Python dictionaries are written with curly brackets, and they have keys and values.

In [124]:
thisdict =	{
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}

# Get the value of the "model" key:
x = thisdict["model"]
print(x)

# There is also a method called get() that will give you the same result:
# Get the value of the "brand" key:
x = thisdict.get("brand")   
print(x)

# Change the "year" to 2018:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict["year"] = 2018

# Print all key names in the dictionary, one by one:
for x in thisdict:
  print('key: ' + x)

# Print all values in the dictionary, one by one:
for x in thisdict:
  print('values: ' + str(thisdict[x]))

# You can also use the values() method to return values of a dictionary:
for x in thisdict.values():
  print('values2: ' + str(x))

# Loop through both keys and values, by using the items() method:
for x, y in thisdict.items():
  print('key: '+ str(x), 'value: ' + str(y))

# Check if "model" is present in the dictionary:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
if "model" in thisdict:
  print("Yes, 'model' is one of the keys in the thisdict dictionary")
if "Ford" in thisdict.values():
  print("Yes, 'Ford' is one of the values in the thisdict dictionary")

# Print the number of items in the dictionary:
print(len(thisdict))

# Adding an item to the dictionary is done by using a new index key and assigning a value to it:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict["color"] = "red"
print(thisdict)

# The pop() method removes the item with the specified key name:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.pop("model")
print(thisdict)

# The popitem() method removes the last inserted item (in versions before 3.7, a random item is removed instead):
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.popitem()
print(thisdict)

# The del keyword removes the item with the specified key name:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
del thisdict["model"]
print(thisdict)

# The del keyword can also delete the dictionary completely:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
del thisdict

# The clear() method empties the dictionary:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
thisdict.clear()
print(thisdict)

# Make a copy of a dictionary with the copy() method:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
mydict = thisdict.copy()
print(mydict)

# Make a copy of a dictionary with the dict() function:
# Another way to make a copy is to use the built-in function dict().
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
mydict = dict(thisdict)
print(mydict)

# It is also possible to use the dict() constructor to make a new dictionary:
thisdict = dict(brand="Ford", model="Mustang", year=1964)
# note that keywords are not string literals
# note the use of equals rather than colon for the assignment
print(thisdict)

Mustang
Ford
key: brand
key: model
key: year
values: Ford
values: Mustang
values: 2018
values2: Ford
values2: Mustang
values2: 2018
key: brand value: Ford
key: model value: Mustang
key: year value: 2018
Yes, 'model' is one of the keys in the thisdict dictionary
Yes, 'Ford' is one of the values in the thisdict dictionary
3
{'brand': 'Ford', 'model': 'Mustang', 'year': 1964, 'color': 'red'}


'Mustang'

{'brand': 'Ford', 'year': 1964}


('year', 1964)

{'brand': 'Ford', 'model': 'Mustang'}
{'brand': 'Ford', 'year': 1964}
{}
{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}
{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}


### Nested Dictionaries
A dictionary can also contain many dictionaries, this is called nested dictionaries.

In [125]:
# Create a dictionary that contain three dictionaries:
myfamily = {
  "child1" : {
    "name" : "Emil",
    "year" : 2004
  },
  "child2" : {
    "name" : "Tobias",
    "year" : 2007
  },
  "child3" : {
    "name" : "Linus",
    "year" : 2011
  }
}

# Create three dictionaries, then create one dictionary that will contain the other three dictionaries:
child1 = {
  "name" : "Emil",
  "year" : 2004
}
child2 = {
  "name" : "Tobias",
  "year" : 2007
}
child3 = {
  "name" : "Linus",
  "year" : 2011
}

myfamily = {
  "child1" : child1,
  "child2" : child2,
  "child3" : child3
}

|Method|	Description|
|-----|------|
|clear()	|Removes all the elements from the dictionary|
|copy()	|Returns a copy of the dictionary|
|fromkeys()	|Returns a dictionary with the specified keys and value|
|get()	|Returns the value of the specified key|
|items()	|Returns a list containing a tuple for each key value pair|
|keys()|	Returns a list containing the dictionary's keys|
|pop()	|Removes the element with the specified key|
|popitem()	|Removes the last inserted key-value pair|
|setdefault()	|Returns the value of the specified key. If the key does not exist: insert |the key, with the specified value|
|update()|	Updates the dictionary with the specified key-value pairs|
|values()	|Returns a list of all the values in the dictionary|

# 11) Python If ... Else
Python supports the usual logical conditions from mathematics:

Equals: a == b <br>
Not Equals: a != b <br>
Less than: a < b <br>
Less than or equal to: a <= b <br>
Greater than: a > b <br>
Greater than or equal to: a >= b <br>
These conditions can be used in several ways, most commonly in "if statements" and loops. 

An "if statement" is written by using the if keyword.

In [131]:
# If statement:
a = 33
b = 200
if b > a:
  print("b is greater than a")


# Indentation
# Python relies on indentation 
# (whitespace at the beginning of a line) to define scope in the code. 
# Other programming languages often use curly-brackets for this purpose.

# a = 33
# b = 200
# if b > a:
# print("b is greater than a") # you will get an error #uncommend for error


# The elif keyword is pythons way of saying 
# "if the previous conditions were not true, then try this condition".
a = 33
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")

# The else keyword catches anything which isn't caught by the preceding conditions.
a = 200
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")

# You can also have an else without the elif:
a = 200
b = 33
if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

# If you have only one statement to execute, you can put it on the same line as the if statement.
if a > b: print("a is greater than b")

# Short Hand If ... Else
# If you have only one statement to execute, one for if, and one for else, you can put it all on the same line:
a = 2
b = 330
print("A") if a > b else print("B")

# One line if else statement, with 3 conditions:
a = 330
b = 330
print("A") if a > b else print("=") if a == b else print("B")


b is greater than a
a and b are equal
a is greater than b
b is not greater than a
a is greater than b
B


### And
The and keyword is a logical operator, and is used to combine conditional statements:

In [133]:
# Test if a is greater than b, AND if c is greater than a:
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

# Test if a is greater than b, OR if a is greater than c:
a = 200
b = 33
c = 500
if a > b or a > c:
  print("At least one of the conditions is True")

# You can have if statements inside if statements, this is called nested if statements.
x = 41
if x > 10:
  print("Above ten,")
  if x > 20:
    print("and also above 20!")
  else:
    print("but not above 20.")
    
# if statements cannot be empty, but if you for some reason have an if statement with no content, put in the pass statement to avoid getting an error.
a = 33
b = 200

if b > a:
  pass

Both conditions are True
At least one of the conditions is True
Above ten,
and also above 20!


# 12) Python While Loops

Python has two primitive loop commands:
- while loops
- for loops

In [141]:
# Print i as long as i is less than 6:
i = 1
while i < 6:
  print('Q1: ' + str(i))
  i += 1
    
# Exit the loop when i is 3:
i = 1
while i < 6:
  print('Q2: ' + str(i))
  if i == 3:
    break
  i += 1

# Continue to the next iteration if i is 3:
# With the continue statement we can stop the current iteration, and continue with the next:
i = 0
while i < 6:
  i += 1
  if i == 3:
    continue
  print('Q3: ' + str(i))

# Print a message once the condition is false:
i = 1
while i < 6:
  print('Q4: ' + str(i))
  i += 1
else:
  print("i is no longer less than 6")

Q1: 1
Q1: 2
Q1: 3
Q1: 4
Q1: 5
Q2: 1
Q2: 2
Q2: 3
Q3: 1
Q3: 2
Q3: 4
Q3: 5
Q3: 6
Q4: 1
Q4: 2
Q4: 3
Q4: 4
Q4: 5
i is no longer less than 6


# 13) Python For Loops
A for loop is used for iterating over a sequence (that is either a list, a tuple, a dictionary, a set, or a string).

In [152]:
# Print each fruit in a fruit list:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  print(x)

# Loop through the letters in the word "banana":
# Even strings are iterable objects, they contain a sequence of characters:
for x in "banana":
  print(x)

# Exit the loop when x is "banana":
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  print(x)
  if x == "banana":
    break

# Exit the loop when x is "banana", but this time the break comes before the print:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  if x == "banana":
    break
  print(x)

# Do not print banana:
fruits = ["apple", "banana", "cherry"]
for x in fruits:
  if x == "banana":
    continue
  print(x)

# Using the range() function:
for x in range(6):
  print(x)

# Using the start parameter:
for x in range(12, 16):
  print(x)

# Increment the sequence with 3 (default is 1):
for x in range(2, 30, 3):
  print(x)

# Print all numbers from 0 to 5, and print a message when the loop has ended:
for x in range(6):
  print(x)
else:
  print("Finally finished!")

# Nested Loops
# A nested loop is a loop inside a loop.
# The "inner loop" will be executed one time for each iteration of the "outer loop":
# Print each adjective for every fruit:
adj = ["red", "big", "tasty"]
fruits = ["apple", "banana", "cherry"]

for x in adj:
  for y in fruits:
    print(x, y)
    
# for loops cannot be empty, but if you for some reason have a for loop with no content, put in the pass statement to avoid getting an error.
for x in [0, 1, 2]:
  pass

apple
banana
cherry
b
a
n
a
n
a
apple
banana
apple
apple
cherry
0
1
2
3
4
5
12
13
14
15
2
5
8
11
14
17
20
23
26
29
0
1
2
3
4
5
Finally finished!
red apple
red banana
red cherry
big apple
big banana
big cherry
tasty apple
tasty banana
tasty cherry


# 14) Python Functions

A function is a block of code which only runs when it is called. <br>
You can pass data, known as parameters, into a function.<br>
A function can return data as a result.<br>

In [168]:
# In Python a function is defined using the def keyword:
def my_function():
  print("Hello from a function")

my_function()

# Information can be passed into functions as arguments.
# Arguments are specified after the function name, inside the parentheses. 
# You can add as many arguments as you want, just separate them with a comma
def my_function(fname):
  print(fname + " Refsnes")

my_function("Emil")
my_function("Tobias")
my_function("Linus")

# This function expects 2 arguments, and gets 2 arguments:
def my_function(fname, lname):
  print(fname + " " + lname)

my_function("Emil", "Refsnes")

# If the number of arguments is unknown, add a * before the parameter name:
def my_function(*kids):
  print("The youngest child is " + kids[2])

my_function("Emil", "Tobias", "Linus")

# You can also send arguments with the key = value syntax.
# This way the order of the arguments does not matter.
def my_function(child3, child2, child1):
  print("The youngest child is " + child3)

my_function(child1 = "Emil", child2 = "Tobias", child3 = "Linus")

# If the number of keyword arguments is unknown, add a double ** before the parameter name:
def my_function(**kid):
  print("His last name is " + kid["lname"])

my_function(fname = "Tobias", lname = "Refsnes")

# Default Parameter Value
def my_function(country = "Norway"):
  print("I am from " + country)

my_function("Sweden")
my_function("India")
my_function()
my_function("Brazil")

# Passing a List as an Argument
# You can send any data types of argument to a function (string, number, list, dictionary etc.), 
# and it will be treated as the same data type inside the function.
def my_function(food):
  for x in food:
    print(x)

fruits = ["apple", "banana", "cherry"]

my_function(fruits)

# To let a function return a value, use the return statement:
def my_function(x):
  return 5 * x

print(my_function(3))
print(my_function(5))
print(my_function(9))

# The pass Statement
# function definitions cannot be empty, 
# but if you for some reason have a function definition with no content, 
# put in the pass statement to avoid getting an error.
def myfunction():
  pass

# Recursion
# Python also accepts function recursion, which means a defined function can call itself.
def tri_recursion(k, run_no):
  if(k > 0):
    result = k + tri_recursion(k - 1, run_no + 1)
    print(str(run_no) + " " + str(result))
  else:
    result = 0
  return result

print("\n\nRecursion Example Results")
tri_recursion(6, 0)

Hello from a function
Emil Refsnes
Tobias Refsnes
Linus Refsnes
Emil Refsnes
The youngest child is Linus
The youngest child is Linus
His last name is Refsnes
I am from Sweden
I am from India
I am from Norway
I am from Brazil
apple
banana
cherry
15
25
45


Recursion Example Results
5 1
4 3
3 6
2 10
1 15
0 21


21

# 15) Python Lambda

A lambda function is a small anonymous function.

A lambda function can take any number of arguments, but can only have one expression.

In [170]:
# A lambda function that adds 10 to the number passed in as an argument, and print the result:
x = lambda a : a + 10
print(x(5))

# A lambda function that multiplies argument a with argument b and print the result:
x = lambda a, b : a * b
print(x(5, 6))

# A lambda function that sums argument a, b, and c and print the result:
x = lambda a, b, c : a + b + c
print(x(5, 6, 2))

15
30
13


### Why Use Lambda Functions?
The power of lambda is better shown when you use them as an anonymous function inside another function.

Say you have a function definition that takes one argument, and that argument will be multiplied with an unknown number:

Use that function definition to make a function that always doubles the number you send in:

In [171]:
def myfunc(n):
  return lambda a : a * n

mydoubler = myfunc(2)

print(mydoubler(11))

22


# 16) Python Arrays
Arrays are used to store multiple values in one single variable

Python does not have built-in support for Arrays, but Python Lists can be used instead.

Note: This page shows you how to use LISTS as ARRAYS, however, to work with arrays in Python you will have to import a library, like the NumPy library.

In [176]:
# Create an array containing car names:
cars = ["Ford", "Volvo", "BMW"]

# Get the value of the first array item:
x = cars[0]
print(x)

# Print each item in the cars array:
for x in cars:
  print(x)

# Add one more element to the cars array:
cars.append("Honda")

# Delete the second element of the cars array:
cars.pop(1)

# Delete the element that has the value "Volvo":
# Note: The list's remove() method only removes the first occurrence of the specified value.
cars.remove("BMW")


Ford
Ford
Volvo
BMW


'Volvo'

| Method	|Description|
|-----------|------------|
|append()	|Adds an element at the end of the list|
|clear()	|Removes all the elements from the list|
|copy()	|Returns a copy of the list|
|count()	|Returns the number of elements with the specified value|
|extend()	|Add the elements of a list (or any iterable), to the end of the current list|
|index()	|Returns the index of the first element with the specified value|
|insert()	|Adds an element at the specified position|
|pop()	|Removes the element at the specified position|
|remove()	|Removes the first item with the specified value|
|reverse()|	Reverses the order of the list|
|sort()	|Sorts the list|

# 17) Python Classes and Objects
Python is an object oriented programming language.

Almost everything in Python is an object, with its properties and methods.

A Class is like an object constructor, or a "blueprint" for creating objects.

In [1]:
# Create a class named MyClass, with a property named x:
class MyClass:
  x = 5

# Create an object named p1, and print the value of x:
p1 = MyClass()
print(p1.x)

5


### The __init__() Function

The examples above are classes and objects in their simplest form, and are not really useful in real life applications.

To understand the meaning of classes we have to understand the built-in __init__() function.

All classes have a function called __init__(), which is always executed when the class is being initiated.

Use the __init__() function to assign values to object properties, or other operations that are necessary to do when the object is being created:

The __init__() function is called automatically every time the class is being used to create a new object.

In [8]:
# Create a class named Person, use the __init__() function to assign values for name and age:
class Person:
  def __init__(self, name, age):
    self.name = name
    self.age = age

p1 = Person("Peter", 36)

print(p1.name)
print(p1.age)

Peter
36


### Object Methods
Objects can also contain methods. Methods in objects are functions that belong to the object.

Let us create a method in the Person class:

In [11]:
# Insert a function that prints a greeting, and execute it on the p1 object:

class Person:
  def __init__(self, name, age):
    self.name = name
    self.age = age

  def myfunc(self):
    print("Hello my name is " + self.name)

p1 = Person("John", 36)
p1.myfunc()

Hello my name is John


### The self Parameter

The self parameter is a reference to the current instance of the class, and is used to access variables that belongs to the class.

It does not have to be named self , you can call it whatever you like, but it has to be the first parameter of any function in the class:

In [17]:
# Use the words mysillyobject and abc instead of self:
class Person:
  def __init__(mysillyobject, name, age):
    mysillyobject.name = name
    mysillyobject.age = age

  def myfunc(abc):
    print("Hello my name is " + abc.name + str(abc.age))

p1 = Person("John", 36)
p1.myfunc()

# Set the age of p1 to 40:
p1.age = 40
p1.myfunc()

# Delete the age property from the p1 object:
del p1.age

# Delete the p1 object:
del p1

# class definitions cannot be empty, but if you for some reason have a class definition with no content, put in the pass statement to avoid getting an error.
lass Person:
  pass

Hello my name is John36
Hello my name is John40


# 18) Python Inheritance

Inheritance allows us to define a class that inherits all the methods and properties from another class.

Parent class is the class being inherited from, also called base class.

Child class is the class that inherits from another class, also called derived class.

In [18]:
# Create a Parent Class
# Any class can be a parent class, so the syntax is the same as creating any other class:
class Person:
  def __init__(self, fname, lname):
    self.firstname = fname
    self.lastname = lname

  def printname(self):
    print(self.firstname, self.lastname)

#Use the Person class to create an object, and then execute the printname method:

x = Person("John", "Doe")
x.printname()

John Doe


In [20]:
# Create a Child Class
# To create a class that inherits the functionality from another class, send the parent class as a parameter when creating the child class:
class Student(Person):
  pass

# Use the Student class to create an object, and then execute the printname method:
x = Student("Mike", "Olsen")
x.printname()

Mike Olsen


### Add the __init__() Function

When you add the __init__() function, the child class will no longer inherit the parent's __init__() function.

Note: The child's __init__() function overrides the inheritance of the parent's __init__() function.

To keep the inheritance of the parent's __init__() function, add a call to the parent's __init__() function:

In [21]:
class Student(Person):
  def __init__(self, fname, lname):
    Person.__init__(self, fname, lname)

### Use the super() Function
Python also has a super() function that will make the child class inherit all the methods and properties from its parent:

In [22]:
class Student(Person):
  def __init__(self, fname, lname):
    super().__init__(fname, lname)

### Add Properties

In [23]:
# Add a year parameter, and pass the correct year when creating objects:
class Student(Person):
  def __init__(self, fname, lname, year):
    super().__init__(fname, lname)
    self.graduationyear = year

x = Student("Mike", "Olsen", 2019)

### Add Methods

In [24]:
# Add a method called welcome to the Student class:
class Student(Person):
  def __init__(self, fname, lname, year):
    super().__init__(fname, lname)
    self.graduationyear = year

  def welcome(self):
    print("Welcome", self.firstname, self.lastname, "to the class of", self.graduationyear)

# 19) Python Iterators

An iterator is an object that contains a countable number of values.

An iterator is an object that can be iterated upon, meaning that you can traverse through all the values.

Technically, in Python, an iterator is an object which implements the iterator protocol, which consist of the methods __iter__() and __next__().

### Iterator vs Iterable

Lists, tuples, dictionaries, and sets are all iterable objects. They are iterable containers which you can get an iterator from.

All these objects have a iter() method which is used to get an iterator:

In [27]:
# Return an iterator from a tuple, and print each value:
mytuple = ("apple", "banana", "cherry")
myit = iter(mytuple)

print(next(myit))
print(next(myit))
print(next(myit))

# Strings are also iterable objects, containing a sequence of characters:
mystr = "banana"
myit = iter(mystr)

print(next(myit))
print(next(myit))
print(next(myit))
print(next(myit))
print(next(myit))
print(next(myit))

# Iterate the values of a tuple:
mytuple = ("apple", "banana", "cherry")
for x in mytuple:
  print(x)

# Iterate the characters of a string:
mystr = "banana"
for x in mystr:
  print(x)

apple
banana
cherry
b
a
n
a
n
a
apple
banana
cherry
b
a
n
a
n
a


### Create an Iterator
To create an object/class as an iterator you have to implement the methods __iter__() and __next__() to your object.

As you have learned in the Python Classes/Objects chapter, all classes have a function called __init__(), which allows you to do some initializing when the object is being created.

The __iter__() method acts similar, you can do operations (initializing etc.), but must always return the iterator object itself.

The __next__() method also allows you to do operations, and must return the next item in the sequence.

In [28]:
class MyNumbers:
  def __iter__(self):
    self.a = 1
    return self

  def __next__(self):
    x = self.a
    self.a += 1
    return x

myclass = MyNumbers()
myiter = iter(myclass)

print(next(myiter))
print(next(myiter))
print(next(myiter))
print(next(myiter))
print(next(myiter))

1
2
3
4
5


## StopIteration
The example above would continue forever if you had enough next() statements, or if it was used in a for loop.

To prevent the iteration to go on forever, we can use the StopIteration statement.

In the __next__() method, we can add a terminating condition to raise an error if the iteration is done a specified number of times:

In [29]:
# Stop after 20 iterations:
class MyNumbers:
  def __iter__(self):
    self.a = 1
    return self

  def __next__(self):
    if self.a <= 20:
      x = self.a
      self.a += 1
      return x
    else:
      raise StopIteration

myclass = MyNumbers()
myiter = iter(myclass)

for x in myiter:
  print(x)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


# 20) Python Scope

A variable is only available from inside the region it is created. This is called scope.

### Local Scope

A variable created inside a function belongs to the local scope of that function, and can only be used inside that function.

In [30]:
# A variable created inside a function is available inside that function:
def myfunc():
  x = 300
  print(x)

myfunc()

300


## Function Inside Function

As explained in the example above, the variable x is not available outside the function, but it is available for any function inside the function:

In [31]:
# The local variable can be accessed from a function within the function:
def myfunc():
  x = 300
  def myinnerfunc():
    print(x)
  myinnerfunc()

myfunc()

300


## Global Scope
A variable created in the main body of the Python code is a global variable and belongs to the global scope.

Global variables are available from within any scope, global and local.

In [33]:
# A variable created outside of a function is global and can be used by anyone:

x = 300

def myfunc():
  print(x)

myfunc()

print(x)

300
300


## Naming Variables
If you operate with the same variable name inside and outside of a function, Python will treat them as two separate variables, one available in the global scope (outside the function) and one available in the local scope (inside the function)

In [34]:
# The function will print the local x, and then the code will print the global x:

x = 300

def myfunc():
  x = 200
  print(x)

myfunc()

print(x)

200
300


## Global Keyword
If you need to create a global variable, but are stuck in the local scope, you can use the global keyword.

The global keyword makes the variable global.

In [35]:
# If you use the global keyword, the variable belongs to the global scope:

def myfunc():
  global x
  x = 300

myfunc()

print(x)

300


In [36]:
# To change the value of a global variable inside a function, refer to the variable by using the global keyword:

x = 300

def myfunc():
  global x
  x = 200

myfunc()

print(x)

200


# 21) Python Modules

What is a Module?
Consider a module to be the same as a code library.

A file containing a set of functions you want to include in your application.

Create a Module
To create a module just save the code you want in a file with the file extension .py:

In [40]:
# Save this code in a file named mymodule.py
def greeting(name):
  print("Hello, " + name)

# Import the module named mymodule, and call the greeting function:
import mymodule
mymodule.greeting("Jonathan")



Hello, Jonathan


## Variables in Module
The module can contain functions, as already described, but also variables of all types (arrays, dictionaries, objects etc):

In [48]:
# Save this code in the file mymodule.py
person1 = {
  "name": "John",
  "age": 36,
  "country": "Norway"
}

# Import the module named mymodule, and access the person1 dictionary:
import mymodule1
a = mymodule1.person1["age"]
print(a)

36


## Re-naming a Module
You can create an alias when you import a module, by using the as keyword:

In [50]:
# Create an alias for mymodule called mx:
import mymodule1 as mx

a = mx.person1["age"]
print(a)

36


## Built-in Modules
There are several built-in modules in Python, which you can import whenever you like.

In [51]:
import platform

x = platform.system()
print(x)

Darwin


## Using the dir() Function
There is a built-in function to list all the function names (or variable names) in a module. The dir() function:

In [53]:
# List all the defined names belonging to the platform module:
import platform

x = dir(platform)
print(x)

x2 = platform.__version__
print(x2)

['DEV_NULL', '_UNIXCONFDIR', '_WIN32_CLIENT_RELEASES', '_WIN32_SERVER_RELEASES', '__builtins__', '__cached__', '__copyright__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', '__version__', '_comparable_version', '_component_re', '_default_architecture', '_dist_try_harder', '_follow_symlinks', '_ironpython26_sys_version_parser', '_ironpython_sys_version_parser', '_java_getprop', '_libc_search', '_linux_distribution', '_lsb_release_version', '_mac_ver_xml', '_node', '_norm_version', '_parse_release_file', '_platform', '_platform_cache', '_pypy_sys_version_parser', '_release_filename', '_release_version', '_supported_dists', '_sys_version', '_sys_version_cache', '_sys_version_parser', '_syscmd_file', '_syscmd_uname', '_syscmd_ver', '_uname_cache', '_ver_output', '_ver_stages', 'architecture', 'collections', 'dist', 'java_ver', 'libc_ver', 'linux_distribution', 'mac_ver', 'machine', 'node', 'os', 'platform', 'popen', 'processor', 'python_branch', 'python_build

## Import From Module

You can choose to import only parts from a module, by using the from keyword.

In [56]:
# Import only the person2 dictionary from the module:
from mymodule2 import person2
print (person2["age"])

46


# 21) Python Datetime

A date in Python is not a data type of its own, but we can import a module named datetime to work with dates as date objects.

In [57]:
# Import the datetime module and display the current date:
import datetime
x = datetime.datetime.now()
print(x)

2020-08-13 12:03:42.738167


### Date Output
When we execute the code from the example above the result will be:

2020-08-13 12:02:52.730743
The date contains year, month, day, hour, minute, second, and microsecond.

The datetime module has many methods to return information about the date object.

Here are a few examples, you will learn more about them later in this chapter:

In [58]:
# Return the year and name of weekday:
import datetime
x = datetime.datetime.now()
print(x.year)
print(x.strftime("%A"))

2020
Thursday


## Creating Date Objects
To create a date, we can use the datetime() class (constructor) of the datetime module.

The datetime() class requires three parameters to create a date: year, month, day.

The datetime() class also takes parameters for time and timezone (hour, minute, second, microsecond, tzone), but they are optional, and has a default value of 0, (None for timezone).

In [59]:
# Create a date object:
import datetime
x = datetime.datetime(2020, 5, 17)
print(x)

2020-05-17 00:00:00


### The strftime() Method
The datetime object has a method for formatting date objects into readable strings.

The method is called strftime(), and takes one parameter, format, to specify the format of the returned string:

In [61]:
# Display the name of the month:

import datetime

x = datetime.datetime(2018, 6, 1)

print(x.strftime("%B"))

June


|Directive|	Description|	Example	|
|-|-|-|
|%a	|Weekday, short version|	Wed	|
|%A	|Weekday, full version	|Wednesday	|
|%w	|Weekday as a number 0-6, 0 is Sunday|	3	|
|%d	|Day of month 01-31|	31	|
|%b	|Month name, short version|	Dec	|
|%B	|Month name, full version	|December	|
|%m	|Month as a number 01-12|	12	|
|%y	|Year, short version, without century	|18	|
|%Y	|Year, full version	|2018	|
|%H	|Hour 00-23	|17	|
|%I	|Hour 00-12|	05	|
|%p	|AM/PM	|PM	|
|%M	|Minute 00-59	|41	|
|%S	|Second 00-59	|08	|
|%f	|Microsecond 000000-999999	|548513	|
|%z	|UTC offset	|+0100	|
|%Z	|Timezone	|CST	|
|%j	|Day number of year |001-366	365	|
|%U	|Week number of year, Sunday as the first day of week, 00-53	|52	|
|%W	|Week number of year, Monday as the first day of week, 00-53	|52	|
|%c	|Local version of date and time	M|on Dec 31 17:41:00 2018	|
|%x	|Local version of date	|12/31/18	|
|%X	|Local version of time	|17:41:00	|
|%%	|A % character	%|

# 22) Python Math

Python has a set of built-in math functions, including an extensive math module, that allows you to perform mathematical tasks on numbers.

### Built-in Math Functions
The min() and max() functions can be used to find the lowest or highest value in an iterable:

In [63]:
#find min and max
x = min(5, 10, 25)
y = max(5, 10, 25)

print(x)
print(y)

#The abs() function returns the absolute (positive) value of the specified number:
x = abs(-7.25)
print(x)

# The pow(x, y) function returns the value of x to the power of y (xy).
# Return the value of 4 to the power of 3 (same as 4 * 4 * 4):
x = pow(4, 3)
print(x)

5
25
7.25
64


### The Math Module
Python has also a built-in module called math, which extends the list of mathematical functions.

When you have imported the math module, you can start using methods and constants of the module.

The math.sqrt() method for example, returns the square root of a number:

To use it, you must import the math module:

In [64]:
import math

x = math.sqrt(64)
print(x)

x = math.ceil(1.4)
y = math.floor(1.4)
print(x) # returns 2
print(y) # returns 1

# The math.pi constant, returns the value of PI (3.14...):
x = math.pi
print(x)

8.0
2
1
3.141592653589793


### Math Methods
|Method	| Description|
|-|-|
|math.acos(x)	|Returns the arc cosine value of x|
|math.acosh(x)	|Returns the hyperbolic arc cosine of x|
|math.asin(x)	|Returns the arc sine of x|
|math.asinh(x)	|Returns the hyperbolic arc sine of x|
|math.atan(x)	|Returns the arc tangent value of x|
|math.atan2(y, x)	|Returns the arc tangent of y/x in radians|
|math.atanh(x)	|Returns the hyperbolic arctangent value of x|
|math.ceil(x)	|Rounds a number upwards to the nearest integer, and returns the result|
|math.comb(n, k)	|Returns the number of ways to choose k items from n items without repetition and order|
|math.copysign(x, y)|	Returns a float consisting of the value of the first parameter and the sign of the second parameter|
|math.cos(x)|	Returns the cosine of x|
|math.cosh(x)|	Returns the hyperbolic cosine of x|
|math.degrees(x)|	Converts an angle from radians to degrees|
|math.dist(p, q)|	Calculates the euclidean distance between two specified points (p and q), where p and q are the coordinates of that point|
|math.erf(x)|	Returns the error function of x|
|math.erfc(x)|	Returns the complementary error function of x|
|math.exp(x)	|Returns the value of Ex, where E is Euler's number (approximately 2.718281...), and x is the number passed to it|
|math.expm1(x)	|Returns the value of Ex - 1, where E is Euler's number (approximately 2.718281...), and x is the number passed to it|
|math.fabs(x)|	Returns the absolute value of a number|
|math.factorial()|	Returns the factorial of a number|
|math.floor(x)|	Rounds a number downwards to the nearest integer, and returns the result|
|math.fmod(x, y)|	Returns the remainder of specified numbers when a number is divided by another number|
|math.frexp()|	Returns the mantissa and the exponent, of a specified value|
|math.fsum(iterable)|	Returns the sum of all items in an iterable (tuples, arrays, lists, etc.)|
|math.gamma(x)|	Returns the gamma value of x|
|math.gcd()	| Returns the highest value that can divide two integers|
|math.hypot()	|Find the Euclidean distance from the origin for n inputs|
|math.isclose()	|Checks whether two values are close, or not|
|math.isfinite(x)|	Checks whether x is a finite number|
|math.isinf(x)|	Check whether x is a positive or negative infinty|
|math.isnan(x)	|Checks whether x is NaN (not a number)|
|math.isqrt(n)	|Returns the nearest integer square root of n|
|math.ldexp(x, i)|	Returns the expression x * 2i where x is mantissa and i is an exponent|
|math.lgamma(x)	|Returns the log gamma value of x|
|math.log(x, base)|	Returns the natural logarithm of a number, or the logarithm of number to base|
|math.log10(x)	|Returns the base-10 logarithm of x|
|math.log1p(x)	|Returns the natural logarithm of 1+x|
|math.log2(x)	|Returns the base-2 logarithm of x|
|math.perm(n, k)	|Returns the number of ways to choose k items from n items with order and without repetition|
|math.pow(x, y)|	Returns the value of x to the power of y|
|math.prod(iterable, *, start=1)|	Returns the product of an iterable (lists, array, tuples, etc.)|
|math.radians(x)	|Converts a degree value (x) to radians|
|math.remainder(x, y)	|Returns the closest value that can make numerator completely divisible by the denominator|
|math.sin(x)	|Returns the sine of x|
|math.sinh(x)	|Returns the hyperbolic sine of x|
|math.sqrt(x)	|Returns the square root of x|
|math.tan(x)	|Returns the tangent of x|
|math.tanh(x)	|Returns the hyperbolic tangent of x|
|math.trunc(x)	|Returns the truncated integer parts of x|

### Math Constants
|Constant	|Description|
|-|-|
|math.e	|Returns Euler's number (2.7182...)|
|math.inf	|Returns a floating-point positive infinity|
|math.nan	|Returns a floating-point NaN (Not a Number) value|
|math.pi	|Returns PI (3.1415...)|
|math.tau	|Returns tau (6.2831...)|

# 23) Python JSON

JSON is a syntax for storing and exchanging data.

JSON is text, written with JavaScript object notation.

Python has a built-in package called json, which can be used to work with JSON data.

In [65]:
# Convert from JSON to Python:

import json

# some JSON:
x =  '{ "name":"John", "age":30, "city":"New York"}'

# parse x:
y = json.loads(x)

# the result is a Python dictionary:
print(y["age"])

30


### Convert from Python to JSON

If you have a Python object, you can convert it into a JSON string by using the json.dumps() method.

In [66]:
# Convert from Python to JSON:
import json

# a Python object (dict):
x = {
  "name": "John",
  "age": 30,
  "city": "New York"
}

# convert into JSON:
y = json.dumps(x)

# the result is a JSON string:
print(y)

{"name": "John", "age": 30, "city": "New York"}


You can convert Python objects of the following types, into JSON strings:

dict <br>
list<br>
tuple<br>
string<br>
int<br>
float<br>
True<br>
False<br>
None

In [67]:
# Convert Python objects into JSON strings, and print the values:

import json

print(json.dumps({"name": "John", "age": 30}))
print(json.dumps(["apple", "bananas"]))
print(json.dumps(("apple", "bananas")))
print(json.dumps("hello"))
print(json.dumps(42))
print(json.dumps(31.76))
print(json.dumps(True))
print(json.dumps(False))
print(json.dumps(None))

{"name": "John", "age": 30}
["apple", "bananas"]
["apple", "bananas"]
"hello"
42
31.76
true
false
null


When you convert from Python to JSON, Python objects are converted into the JSON (JavaScript) equivalent:

|Python	|JSON|
|-|-|
|dict	|Object|
|list	|Array|
|tuple	|Array|
|str	|String|
|int	|Number|
|float	|Number|
|True	|true|
|False	|false|
|None	|null|

In [68]:
# Convert a Python object containing all the legal data types:

import json

x = {
  "name": "John",
  "age": 30,
  "married": True,
  "divorced": False,
  "children": ("Ann","Billy"),
  "pets": None,
  "cars": [
    {"model": "BMW 230", "mpg": 27.5},
    {"model": "Ford Edge", "mpg": 24.1}
  ]
}

print(json.dumps(x))

{"name": "John", "age": 30, "married": true, "divorced": false, "children": ["Ann", "Billy"], "pets": null, "cars": [{"model": "BMW 230", "mpg": 27.5}, {"model": "Ford Edge", "mpg": 24.1}]}


### Format the Result

The example above prints a JSON string, but it is not very easy to read, with no indentations and line breaks.

The json.dumps() method has parameters to make it easier to read the result:

In [69]:
# Use the indent parameter to define the numbers of indents:

json.dumps(x, indent=4)

'{\n    "name": "John",\n    "age": 30,\n    "married": true,\n    "divorced": false,\n    "children": [\n        "Ann",\n        "Billy"\n    ],\n    "pets": null,\n    "cars": [\n        {\n            "model": "BMW 230",\n            "mpg": 27.5\n        },\n        {\n            "model": "Ford Edge",\n            "mpg": 24.1\n        }\n    ]\n}'

In [70]:
# You can also define the separators, default value is (", ", ": "), 
# which means using a comma and a space to separate each object, 
# and a colon and a space to separate keys from values:

json.dumps(x, indent=4, separators=(". ", " = "))

'{\n    "name" = "John". \n    "age" = 30. \n    "married" = true. \n    "divorced" = false. \n    "children" = [\n        "Ann". \n        "Billy"\n    ]. \n    "pets" = null. \n    "cars" = [\n        {\n            "model" = "BMW 230". \n            "mpg" = 27.5\n        }. \n        {\n            "model" = "Ford Edge". \n            "mpg" = 24.1\n        }\n    ]\n}'

### Order the Result
The json.dumps() method has parameters to order the keys in the result:

In [71]:
# Use the sort_keys parameter to specify if the result should be sorted or not:

json.dumps(x, indent=4, sort_keys=True)

'{\n    "age": 30,\n    "cars": [\n        {\n            "model": "BMW 230",\n            "mpg": 27.5\n        },\n        {\n            "model": "Ford Edge",\n            "mpg": 24.1\n        }\n    ],\n    "children": [\n        "Ann",\n        "Billy"\n    ],\n    "divorced": false,\n    "married": true,\n    "name": "John",\n    "pets": null\n}'

# 24) Python RegEx

A RegEx, or Regular Expression, is a sequence of characters that forms a search pattern.

RegEx can be used to check if a string contains the specified search pattern.

### RegEx Module
Python has a built-in package called re, which can be used to work with Regular Expressions.

Import the re module:

In [76]:
# Search the string to see if it starts with "The" and ends with "Spain":

import re

txt = "The rain in Spain"
x = re.search("^The.*Spain$", txt)
print(x)
print(bool(x))

<re.Match object; span=(0, 17), match='The rain in Spain'>
True


### The findall() Function
The findall() function returns a list containing all matches.

In [78]:
import re
# Print a list of all matches:
txt = "The rain in Spain"
x = re.findall("ai", txt)
print(x)

# Return an empty list if no match was found:
txt = "The rain in Spain"
x = re.findall("Portugal", txt)
print(x)

['ai', 'ai']
[]


### The search() Function
The search() function searches the string for a match, and returns a Match object if there is a match.

If there is more than one match, only the first occurrence of the match will be returned:

In [80]:
# Search for the first white-space character in the string:
import re
b
txt = "The rain in Spain"
x = re.search("\s", txt)

print("The first white-space character is located in position:", x.start())

The first white-space character is located in position: 3


### The split() Function
The split() function returns a list where the string has been split at each match:

In [82]:
import re

# Split at each white-space character:
txt = "The rain in Spain"
x = re.split("\s", txt)
print(x)

# You can control the number of occurrences by specifying the maxsplit parameter:
# Split the string only at the first occurrence:
txt = "The rain in Spain"
x = re.split("\s", txt, 1)
print(x)

['The', 'rain', 'in', 'Spain']
['The', 'rain in Spain']


### The sub() Function
The sub() function replaces the matches with the text of your choice:

In [84]:
# Replace every white-space character with the number 9:
import re

txt = "The rain in Spain"
x = re.sub("\s", "9", txt)
print(x)

# Replace the first 2 occurrences:
txt = "The rain in Spain"
x = re.sub("\s", "9", txt, 2)
print(x)

The9rain9in9Spain
The9rain9in Spain


### Match Object
A Match Object is an object containing information about the search and the result.

The Match object has properties and methods used to retrieve information about the search, and the result:

.span() returns a tuple containing the start-, and end positions of the match.<br>
.string returns the string passed into the function <br>
.group() returns the part of the string where there was a match

In [86]:
# Do a search that will return a Match Object:

import re

txt = "The rain in Spain"
x = re.search("ai", txt)
print(x) #this will print an object

# The regular expression looks for any words that starts with an upper case "S":
txt = "The rain in Spain"
x = re.search(r"\bS\w+", txt)
print(x.span())

# Print the string passed into the function:
txt = "The rain in Spain"
x = re.search(r"\bS\w+", txt)
print(x.string)

# The regular expression looks for any words that starts with an upper case "S":
txt = "The rain in Spain"
x = re.search(r"\bS\w+", txt)
print(x.group())

<re.Match object; span=(5, 7), match='ai'>
(12, 17)
The rain in Spain
Spain


### RegEx Functions
The re module offers a set of functions that allows us to search a string for a match:

| Function	| Description| 
| -| -| 
| findall	| Returns a list containing all matches| 
| search	| Returns a Match object if there is a match anywhere in the string| 
| split	| Returns a list where the string has been split at each match| 
| sub	| Replaces one or many matches with a string| 

### Metacharacters
Metacharacters are characters with a special meaning:

| Character	| Description| 	Example	| 
| -| -| -| 
| []	| A set of characters	| "[a-m]"	| 
| \	| Signals a special sequence (can also be used to escape special characters)| 	"\d" | | .	| Any character (except newline character)	| "he..o"	| 
| ^	| Starts with	| "^hello"	| 
| \$| Ends with| "world$"	| 
| *	| Zero or more occurrences	| "aix*"	| 
| +	| One or more occurrences| 	"aix+"	| 
| {}	| Exactly the specified number of occurrences| 	"al{2}"	| 
|	'| 'Either or	"falls|stays"	| 
| ()	| Capture and group| - | 

### Special Sequences
A special sequence is a \ followed by one of the characters in the list below, and has a special meaning:

| Character	|Description	|Example	|
|-|-|-|
|\A	|Returns a match if the specified characters are at the beginning of the string	|"\AThe"	|
\b	|Returns a match where the specified characters are at the beginning or at the end of a word (the "r" in the beginning is making sure that the string is being treated as a "raw string")	|r"\bain" r"ain\b"	|
|\B	|Returns a match where the specified characters are present, but NOT at the beginning (or at the end) of a word (the "r" in the beginning is making sure that the string is being treated as a "raw string")	|r"\Bain" r"ain\B"	|
|\d	|Returns a match where the string contains digits (numbers from 0-9)|	"\d"	|
|\D	|Returns a match where the string DOES NOT contain digits	|"\D"	|
|\s	|Returns a match where the string contains a white space character	|"\s"	|
|\S	|Returns a match where the string DOES NOT contain a white space character	|"\S"	|
|\w	|Returns a match where the string contains any word characters (characters from a to Z, digits from 0-9, and the underscore _ character)|	"\w"	|
|\W	|Returns a match where the string DOES NOT contain any word characters	|"\W"	|
|\Z	|Returns a match if the specified characters are at the end of the string	|"Spain\Z"|

### Sets
A set is a set of characters inside a pair of square brackets [] with a special meaning:

|Set	|Description	|
|-|-|
|[arn]	|Returns a match where one of the specified characters (a, r, or n) are present	|
|[a-n]	|Returns a match for any lower case character, alphabetically between a and n	|
|[^arn]	|Returns a match for any character EXCEPT a, r, and n	|
|[0123]	|Returns a match where any of the specified digits (0, 1, 2, or 3) are present	|
|[0-9]	|Returns a match for any digit between 0 and 9	|
|[0-5][0-9]|	Returns a match for any two-digit numbers from 00 and 59	|
|[a-zA-Z]	|Returns a match for any character alphabetically between a and z, lower case OR upper case	|
|[+]	|In sets, +, *, ., |, (), $,{} has no special meaning, so [+] means: return a match for any + character in the string|

# 25) Python Try Except
The try block lets you test a block of code for errors.

The except block lets you handle the error.

The finally block lets you execute code, regardless of the result of the try- and except blocks.

### Exception Handling
When an error occurs, or exception as we call it, Python will normally stop and generate an error message.

These exceptions can be handled using the try statement:

In [99]:
# The try block will generate an exception, because gggg is not defined:
try:
  print(gggg)
except:
  print("An exception occurred")

# Print one message if the try block raises a NameError and another for other errors:
try:
  print(ggggg2)
except NameError:
  print("Variable ggggg2 is not defined")
except:
  print("Something else went wrong")

# In this example, the try block does not generate any error:
try:
  print("Hello")
except:
  print("Something went wrong")
else:
  print("Nothing went wrong")

# The finally block, 
# if specified, will be executed regardless if the try block raises an error or not.
try:
  print(x)
except:
  print("Something went wrong")
finally:
  print("The 'try except' is finished")

# This can be useful to close objects and clean up resources:
# Try to open and write to a file that is not writable:
try:
  af = open("demofile.txt")
  af.write("Lorum Ipsum")
except:
  print("Something went wrong when writing to the file")

# Raise an error and stop the program if x is lower than 0:
x = -1
if x < 0:
  raise Exception("Sorry, no numbers below zero")

An exception occurred
Variable ggggg2 is not defined
Hello
Nothing went wrong
-1
The 'try except' is finished
Something went wrong when writing to the file


Exception: Sorry, no numbers below zero

In [100]:
# Raise a TypeError if x is not an integer:
x = "hello"
if not type(x) is int:
  raise TypeError("Only integers are allowed")

TypeError: Only integers are allowed

# 26) Python User Input

Python allows for user input.

That means we are able to ask the user for input.

The method is a bit different in Python 3.6 than Python 2.7.

Python 3.6 uses the input() method.

Python 2.7 uses the raw_input() method.

The following example asks for the username, and when you entered the username, it gets printed on the screen:

In [101]:
username = input("Enter username:")
print("Username is: " + username)

Enter username:abc
Username is: abc


# 27) Python File Handling
File handling is an important part of any web application.

Python has several functions for creating, reading, updating, and deleting files.

### File Handling
The key function for working with files in Python is the open() function.

The open() function takes two parameters; filename, and mode.

There are four different methods (modes) for opening a file:

"r" - Read - Default value. Opens a file for reading, error if the file does not exist

"a" - Append - Opens a file for appending, creates the file if it does not exist

"w" - Write - Opens a file for writing, creates the file if it does not exist

"x" - Create - Creates the specified file, returns an error if the file exists

In addition you can specify if the file should be handled as binary or text mode

"t" - Text - Default value. Text mode

"b" - Binary - Binary mode (e.g. images)

In [115]:
# To open a file for reading it is enough to specify the name of the file:
f = open("demofile.txt")

# The code above is the same as:
f = open("demofile.txt", "rt")
print(f.read())

# Open a file on a different location:
# f = open("D:\\myfiles\welcome.txt", "r")

# Return the 5 first characters of the file:
f = open("demofile.txt", "r")
print(f.read(5))

# Read one line of the file:
f = open("demofile.txt", "r")
print(f.readline())

# Read two lines of the file:
f = open("demofile.txt", "r")
print(f.readline())
print(f.readline())

# Loop through the file line by line:
f = open("demofile.txt", "r")
for x in f:
  print(x)

# Close the file when you are finish with it:
f = open("demofile.txt", "r")
print(f.readline())
f.close()

Hello! Welcome to demofile.txt
This file is for testing purposes.
Good Luck!
Hello
Hello! Welcome to demofile.txt

Hello! Welcome to demofile.txt

This file is for testing purposes.

Hello! Welcome to demofile.txt

This file is for testing purposes.

Good Luck!
Hello! Welcome to demofile.txt



# 27.1) Python File Write

Write to an Existing File <br>
To write to an existing file, you must add a parameter to the open() function:

"a" - Append - will append to the end of the file

"w" - Write - will overwrite any existing content

In [117]:
# Open the file "demofile2.txt" and append content to the file:

f = open("demofile.txt", "a")
f.write("Now the file has more content!")
f.close()

#open and read the file after the appending:
f = open("demofile.txt", "r")
print(f.read())

# Open the file "demofile3.txt" and overwrite the content:
f = open("demofile.txt", "w")
f.write("Woops! I have deleted the content!")
f.close()

#open and read the file after the appending:
f = open("demofile.txt", "r")
print(f.read())

Hello! Welcome to demofile.txt
This file is for testing purposes.
Good Luck!Now the file has more content!Now the file has more content!
Woops! I have deleted the content!


## Create a New File
To create a new file in Python, use the open() method, with one of the following parameters:

"x" - Create - will create a file, returns an error if the file exist

"a" - Append - will create a file if the specified file does not exist

"w" - Write - will create a file if the specified file does not exist

In [123]:
# Create a file called "myfile.txt":
f = open("myfile.txt", "x")

# Create a new file if it does not exist:
f = open("myfile.txt", "w")

## Delete a File
To delete a file, you must import the OS module, and run its os.remove() function:

In [127]:
# Remove the file "myfile.txt":
import os
# Check if File exist:
# To avoid getting an error, 
# you might want to check if the file exists before you try to delete it:
# Check if file exists, then delete it:
if os.path.exists("myfile.txt"):
  os.remove("myfile.txt")
  print('file removed')
else:
  print("The file does not exist")

# Remove the folder "myfolder":
import os
# os.rmdir("myfolder") #UNCOMMEND TO RUN

The file does not exist
